include/hlslib/common/OpenCL.h

```diff
@@ -499,6 +502,97 @@ class Buffer {
   Buffer(Context &context, size_t nElements)
       : Buffer(context, MemoryBank::unspecified, nElements) {}
 
+#ifdef HLSLIB_XILINX
+  /// Allocate DDR or HBM but don't perform any transfers.
+  Buffer(Context &context, StorageType storageType, uint8_t bankIndex, size_t nElements)
+      : context_(&context), nElements_(nElements) {
+#ifndef HLSLIB_SIMULATE_OPENCL
+
+    cl_mem_flags flags;
+    switch (access) {
+      case Access::read:
+        flags = CL_MEM_READ_ONLY;
+        break;
+      case Access::write:
+        flags = CL_MEM_WRITE_ONLY;
+        break;
+      case Access::readWrite:
+        flags = CL_MEM_READ_WRITE;
+        break;
+    }
+
+    void *hostPtr = nullptr;
+    ExtendedMemoryPointer extendedHostPointer = CreateExtendedPointer(
+      nullptr, storageType, bankIndex
+    );
+    hostPtr = &extendedHostPointer;
+    flags |= kXilinxMemPointer;
+
+    cl_int errorCode;
+    {
+      std::lock_guard<std::mutex> lock(context_->memcopyMutex());
+      devicePtr_ = cl::Buffer(context_->context(), flags,
+                              sizeof(T) * nElements_, hostPtr, &errorCode);
+    }
+
+    if (errorCode != CL_SUCCESS) {
+      ThrowRuntimeError("Failed to initialize device memory.");
+      return;
+    }
+#else
+    devicePtr_ = std::make_unique<T[]>(nElements_);
+#endif
+  }
+
+ /// Allocate on HBM or DDR and copy to device
+  template <typename IteratorType, typename = typename std::enable_if<
+                                       IsIteratorOfType<IteratorType, T>() &&
+                                       IsRandomAccess<IteratorType>()>::type>
+  Buffer(Context &context, StorageType storageType, uint8_t bankIndex, IteratorType begin,
+         IteratorType end)
+      : context_(&context), nElements_(std::distance(begin, end)) {
+#ifndef HLSLIB_SIMULATE_OPENCL
+
+    void *hostPtr = nullptr;
+    cl_mem_flags flags;
+
+    switch (access) {
+      case Access::read:
+        flags = CL_MEM_READ_ONLY;
+        break;
+      case Access::write:
+        flags = CL_MEM_WRITE_ONLY;
+        break;
+      case Access::readWrite:
+        flags = CL_MEM_READ_WRITE;
+        break;
+    }
+
+    hostPtr = const_cast<T *>(&(*begin));
+    flags |= CL_MEM_USE_HOST_PTR;
+    // Allow specifying memory bank
+    ExtendedMemoryPointer extendedHostPointer = CreateExtendedPointer(
+      hostPtr, storageType, bankIndex
+    );
+    hostPtr = &extendedHostPointer;
+    flags |= kXilinxMemPointer;
```

Comment:
Same as above: Unify this with the other function so there's only one, where one function translates the arguments into the format of the other